### DISCUSSION TITLE
Q&A

### DISCUSSION THREAD

**Discussion by MyHomeMyData** [MyHomeMyData/E3onCAN#1] [Q&A]
Place questions here.

Fragen gerne hier stellen.

**Comment by M4n197** [comment]
Hallo,
ich hätte da mal eine Frage bzgl. des E380:
Wie ist: `0x258: E3ComplexType(8, "GridEnergy", [E3Int16(2, "ImpCumL", scale=1.0, signed=False), E3Int16(2, "ImpCumT", scale=1.0, signed=False), E3Int16(2, "ExpCumL", scale=1.0, signed=False), E3Int16(2, "ExpCumT", scale=1.0, signed=False),]),`
aus  E3onCANdatapointsE380.py zu verstehen? Bezogene und eingespeiste Energie versetehe ich, aber auf "L" und "T" konnte ich mir noch keinen Reim machen. Nutzt das so schon jemand und hat plausible Werte?

Bisher nutze ich nur den Bezugswert aus `0x25C`, welcher definitiv korrekt ist. (Vergleichswerte lefert die Gridbox)
Gemäß dem Bild aus dem [Viessmann-Forum](https://www.viessmann-community.com/t5/Konnektivitaet/CAN-Bus-Home-Automation-E3-Generation-lokal-und-kostenlos/td-p/356066/page/6?lightbox-message-images-365819=23629i69F0250DE67E181B)  habe ich mit eigenen Scale und Offset die ersten 4 Bytes von `0x258` an den Wert von `0x25C` sehr genau annähern können. Sowohl Offset als auch Scale sind dabei aber ziemlich abstrus..

**Reply by MyHomeMyData** [reply]
Für die Werte aus 0x258 habe ich auch noch keine klare Interpretation. Generell scheinen das Impulszähler zu sein. Der erste Wert erhöht sich - glaube ich - immer, wenn die LED am E380 blinkt, also wenn 1/1000 kWh geflossen ist.

Ich habe mich an dieser Übersicht orientiert. 0x258 scheint der letzten Zeile in der Tabelle zu entsprechen:
![image](https://github.com/MyHomeMyData/E3onCAN/assets/144950531/1be85053-5980-402f-87db-8a9fdb75066e)

**Reply by M4n197** [reply]
Stimmt, diese Info gibt es ja auch. T bezeichnet hierbei wohl den Tarif (T1, T2).  Es sind damit aber nur 2 Werte und nicht 4, was sich auch mit meiner Beobachtung deckt. Ea+(...) und Ea-(...) sind jeweils die Energien aller Phasen und **beider** Tarife.

**Reply by MyHomeMyData** [reply]
Ja, das würde Sinn machen. Ich habe es als je 2 Werte gelesen.

Wenn es zwei 32 Bit Werte sind, müsste es ja einen passenden Überlauf von den unteren 16 auf die oberen 16 Bit geben. Hast Du schon so einen Überlauf beobachtet?

**Reply by M4n197** [reply]
Ich habe nur sporadisch mal Rohwerte tabelliert, da fehlt jedoch exakt dieser Moment.
Seit mehreren Wochen logge ich aber `0x258` mit meiner Umrechnung und die Deckung ist gut:
![grafik](https://github.com/MyHomeMyData/E3onCAN/assets/142239255/7a5fb42c-281d-4941-9224-4b24bc4efc61)
Grün (`0x25C`) und Gelb (`0x258`, untere 4 Bytes mit Skalierungsfaktor: _0,000007810_ und Offset: _9303,106923_)

Trotz der scheinbar sehr guten Annäherung, liefert diese Umrechnung jedoch für ältere (manuell gesicherte) Werte negative Energien und zwar mit deutlichem Fehler von ~100kWh.

**Reply by MyHomeMyData** [reply]
Ich habe jetzt einen 32-Bit-Überlauf beobachtet. Auch Deine Daten lassen sich nur als 32-Bit-Werte interpretieren.
Werde den Codec gleich auf 2 x 32-Bit ändern.

**Reply by MyHomeMyData** [reply]
@M4n197 Ich habe jetzt 0x258 mit den kumulierten Werten von 0x25A und den Werten des Smartmeters (ISKRA MT631) verglichen. Bisher nur für Netzbezug, Einspeisen is grade nicht.

Ergebnis: Wenn ich einen passenden Offset wähle und die Werte von 0x258 mit 0,5 multipliziere, stimmen alle drei Werte überein (Einheit Wh). Abweichungen sind kleiner als 1 Promille. **Demnach entspricht jeder Tick auf 0x258 0,5 Wh**.

Bist Du sicher, dass Deine Berechnung der kWh aus 0x25A stimmt?

**Reply by M4n197** [reply]
`0x25A` (Leistung) nutze ich nicht zur Ermittlung der Energie.  
Da ich keinen zweiten Zähler habe, kann ich nur die Werte von der Gridbox und `0x25C` zum Vergleichen nutzen.  Diese stimmen überein, wenn ich die Umrechnung von oben ansetze. 
War dein Test eine Stichprobe oder hast du Zeitverläufe übereinander gelegt?

Ich bin zunächst von Ws ausgegangen, aber bisher konnte kein realitätsbezogener Umrechnungsfaktor die Kurven so gut annähern.
Leider hat sich vor wenigen Tagen sogar ein "Knick" ergeben, welcher eine Korrektur der Umrechnung erforderlich gemacht hat:
![grafik](https://github.com/MyHomeMyData/E3onCAN/assets/142239255/79e482c6-871c-4ded-896d-31f41ac1a713)
Ich kann mir keinen Fehler vorstellen, der solche Ergebnisse verursacht. `¯\_(ツ)_/¯`
Vielleicht bekomme ich ja Infos vom Hersteller...

**Reply by MyHomeMyData** [reply]
Für den Vergleich 0x258 mit 0x25A kumuliert ist es ein Zeitverlauf, mit dem Smartmeter eine Stichprobe bei 4300 Wh Verbrauch.

Die 0,5 Wh pro Tick passen ja auch zur Doku des E380. Da steht 1000 Pulse pro kWh, also 1 Wh für einen Zyklus LED an, LED aus. Wenn 0x258 die Flanken zählt, sind es pro Inkrement 0,5 Wh.

![image](https://github.com/MyHomeMyData/E3onCAN/assets/144950531/49413606-6cc7-446a-9351-3b83f50e9514)

Ich hatte auch ein paar Monate eine Gridbox in Betrieb. Hatte den Eindruck, die Daten waren nicht sonderlich genau.

**Reply by M4n197** [reply]
Der Zeitverlauf in der Gridbox ist leider viel zu grob. Die Statistik ist bei mir jedoch plausibel und deckt sich mit den Daten vom E380 und VX3, die ja die Quelle dafür bilden. Bis das Smartmeter kommt, habe ich leider keinen Vergleich.

Stimmt deine Umrechnung auch nach mehreren Tagen noch oder wächst der Fehler langsam?
`0x25C` liest du gar nicht mit?

**Reply by MyHomeMyData** [reply]
Nein, die 0x25C lese ich nicht mit, die Daten hole ich vom SmartMeter. Das muss ich am Ende bezahlen, also ist das die Referenz.

Habe jetzt nochmal die 0x258 ausgelesen und verglichen. Verbrauchswert für 5.11. 0:00 bis 10.11. 16:10 sind:
Smartmeter: 79,4066 kWh
E380 0x258: 79,4825 kWh (Umrechnungsfaktor 0,5)

Differenz fast genau 1 Promille.

**Comment by wotandamm** [comment]
Ich versuche gerade meine Kombination VX3 und E380 auszulesen. Ich mach das mit dem aktuellen Wert von 0x258 aus
E3onCANdatapointsE380.py:
0x258: E3ComplexType(8, "GridEnergy", [E3Int32(4, "ImportCumulated", scale=1.0, signed=False), E3Int32(4, "ExportCumulated", scale=1.0, signed=False)])
da kommt folgendes:
{'ImportCumulated': 1211856412.0, 'ExportCumulated': 1244228589.0}
am E380 lese ich auf dem Display:
IMP 190
EXP 2775
Wie passt das zusammen ? Der faktor 0,5 reicht da nicht.

**Reply by MyHomeMyData** [reply]
Hast Du Dir die Tabelle angeschaut? 0x258 ist nach meinem Verständnis die letzte Zeile:
![image](https://github.com/MyHomeMyData/E3onCAN/assets/144950531/159fa6bf-bd8a-4062-96d2-948bc9829948)

Diese Daten werden demnach auf dem Display nicht angezeigt. Ein Abgleich mit den Displaywerten erscheint mir deshalb nicht sinnvoll.
Meine Interpretation ist: Der passende Zähler in 0x258 zählt immer eins hoch, wenn die LED den Zustand wechselt => LED-aus-LED-an entspricht 1/1000 kWh und 2 Impulsen.
Ich kann mich täuschen. Aber es passt sehr gut zu meinem SmartMeter.

**Reply by M4n197** [reply]
![grafik](https://github.com/MyHomeMyData/E3onCAN/assets/142239255/8e6e483e-9225-4bf7-ad04-fa1dbff85826)
@wotandamm  wenn du nicht die Tariffunktion nutzt, also nie auf `T2` wechselst, sollten die Werte vom Display schon mit denen der Schnittstelle übereinstimmen. `0x25C` liefert nur die importierte Energie, ohne Offset und mit Skalierungsfaktor 0,01 kWh.  
Für `0x258` musst du je einen Offset pro Wert ermitteln und dann noch eine Skalierung vornehmen. Ermittle doch mal den Offset anhand der Displaywerte und probiere es mit dem Faktor 0,5. Die importierte Energie von `0x25C` und `0x258` sollten dann übereinstimmen.

**Reply by wotandamm** [reply]
@M4n197 
Wie liest Du 0x25C aus?
In der Datei E3onCANdatapointsE380.py ist er nicht vorhanden:

dataIdentifiersE380 = {
    0x250: E3ComplexType(8, "GridActivePower", [E3Int16(2, "L1", scale=1.0, signed=True), E3Int16(2, "L2", scale=1.0, signed=True), E3Int16(2, "L3", scale=1.0, signed=True), E3Int16(2, "Total", scale=1.0, signed=True)]),
    0x252: E3ComplexType(8, "GridReactivePower", [E3Int16(2, "L1", scale=1.0, signed=True), E3Int16(2, "L2", scale=1.0, signed=True), E3Int16(2, "L3", scale=1.0, signed=True), E3Int16(2, "Total", scale=1.0, signed=True)]),
    0x254: E3ComplexType(8, "GridCurrent", [E3Int16(2, "L1", scale=1.0, signed=True), E3Int16(2, "L2", scale=1.0, signed=True), E3Int16(2, "L3", scale=1.0, signed=True), E3cosPhi(2, "cosPhi", scale=100.0, signed=True)]),
    0x256: E3ComplexType(8, "GridVoltage", [E3Int16(2, "L1", scale=1.0, signed=False), E3Int16(2, "L2", scale=1.0, signed=False), E3Int16(2, "L3", scale=1.0, signed=False), E3Int16(2, "Frequency", scale=100.0, signed=False)]),
    0x258: E3ComplexType(8, "GridEnergy", [E3Int32(4, "ImportCumulated", scale=1.0, signed=False), E3Int32(4, "ExportCumulated", scale=1.0, signed=False)]),
    0x25A: E3ComplexType(8, "GridPower", [E3Int32(4, "ActivePower", scale=10.0, signed=True), E3Int32(4, "ReactivePower", scale=10.0, signed=True)]),
}
@MyHomeMyData:
Der Zähler und 0x258 erhöht sich bei mir laufend, ca. im 0,5 Sekunden Intervall und nicht wenn die Leuchtdiode blinkt.
Ist es möglich das da was falsch eingestellt wurde?
Ich suche schon läner eine Bedienungsanleitung für den E380.

**Reply by MyHomeMyData** [reply]
@wotandamm Wie gesagt, wirklich sicher bin ich bei 0x258 auch nicht. Mein E380 verhält sich halt wie beschrieben.

Habe [hier](https://www.viessmann-community.com/t5/Waermepumpe-Hybridsysteme/Vitocal250a-mit-PV-Anlage-verbinden/td-p/309233) noch einen Post zum Thema gefunden, Da ist auch die Anleitung zum E380 als PDF enthalten.

**Reply by M4n197** [reply]
@wotandamm probiere es mal mit dieser neuen Zeile:
`0x25C: E3ComplexType(4, "GridEnergy", [E3Int32(4, "ImportCumulated", scale=0.01, signed=False)])`

**Reply by MyHomeMyData** [reply]
Ich habe den main Branch aktualisiert:
* Umrechnungsfaktor 2000 für 0x258 => Einheit kWh
* 0x25C ergänzt mit Umrechnungsfaktor 100 => Einheit kWh

0x258 hat vermutlich einen beliebigen Offset.

Anmerkung: Die Formel für die Umrechnung lautet: `value = raw-value / factor`

Dadurch ist es konsistent mit der Implementierung bei open3e.

**Reply by wotandamm** [reply]
Ich komme mit dem Offset bei 0x258 nicht klar.
Displaywerte: IMP 308,84  EXP 2776,0 
0x258 : IMP 608921.967   EXP  622114.7065
Wie wird der Offset berechnet?

**Reply by MyHomeMyData** [reply]
@wotandamm Leider habe ich bisher keine Infos dazu gefunden, wie das Gerät die Absolutwerte von 0x258 berechnet bzw. initial setzt. Deshalb kann ich die Frage zur Offsetberechnung nicht beantworten. Sorry.

**Reply by M4n197** [reply]
@wotandamm Wenn der Umrechnungsfaktor (2000) bei dir funktioniert wie bei @MyHomeMyData, dann wie folgt:
Displaywert = 0x258 - Offset -> Offfset = 0x258 - Displaywert
Für IMP wäre der Offset also 608921,967 - 308,84 = 608613,127.

Ich bin neugierig, ob das bei dir klappt oder ob du meine Situation vorfindest.


Stimmen 0x25C und Displaywert IMP bei dir überein?

**Reply by wotandamm** [reply]
@M4n197  Funktioniert nicht. Der Offset liegt heute für IMP bei 609824,4795.
Der 0x25C Wert entspricht dem Displaywert.

**Reply by wotandamm** [reply]
@M4n197 Ich habe heute mal eine Messreihe mit Iobroker History angelegt.
Bei mir kann ich jetzt IMP berechnen:
IMP(Display) = ImportCumulated/16 - 37748,74
Ich werde da morgen abend mal überprüfen.

**Reply by M4n197** [reply]
Ich bin gespannt. 
Bei mir kam es erneut zu einem "Knick" in den Daten von `0x258` ...

![grafik](https://github.com/MyHomeMyData/E3onCAN/assets/142239255/c6c4266a-c38b-4a37-878a-95b2c40edcb4)

**Reply by wotandamm** [reply]
@M4n197 
Der Wert vom berechneten Imp ist weiterhin gleich dem ausgelesen, knapp zwei Tage jetzt.
Expcum kann ich noch nicht berechnen, zu wenig Sonne.

**Reply by M4n197** [reply]
@MyHomeMyData  & @wotandamm 
Vor zwei Tagen haben ich endlich herausgefunden, was es mit  `0x258` auf sich hat... 
Es handelt sich einfach um zwei Floats! 

Wer es mal schnell prüfen will, kann https://www.h-schmidt.net/FloatConverter/IEEE754de.html ausprobieren. 
Es ist keine besondere Skalierung und auch kein Offset vorhanden.  
Die Werte entsprechen einfach den Energien in Wh. ¯\_(ツ)_/¯

Das hätte ich auch eher mal probieren sollen, aber da sonst überall Integer genutzt werden, war ich wohl "betriebsblind". (>ლ)

**Reply by wotandamm** [reply]
super, hinter her ist man immer schlauer ;-). 
Baut das jemand ein oder muss die Umwandlung selbst gemacht werden?

**Reply by MyHomeMyData** [reply]
@M4n197 Vielen Dank für den Hinweis. Bei mir stimmt mit der Float16-Konvertierung jetzt endlich der dekodierte Wert mit dem im Display angezeigten Wert überein. Habe es testweise im Branch [develop](https://github.com/MyHomeMyData/E3onCAN/tree/develop) eingebaut, Einheit für 0x258 ist weiterhin kWh (Umrechnungsfaktor 1000).

@M4n197 @wotandamm Bitte 0x258 mal ausprobieren.
Für 0x254 (GridCurrent) habe ich auf Float16 umgestellt. Die Werte sehen bei mir immer noch nicht wirklich sinnvoll aus, aber zumindest weniger unsinnig als mit Int16. Könnt ihr das bitte auch mal anschauen? Vielleicht habt ihr eine Idee, was hier das korrekte Format sein könnte. Mit Float32 (könnte ja Wirk- und Blindstrom sein) und Float64 habe ich auch getestet. Da kommt jeweils völliger Blödsinn raus.

**Reply by M4n197** [reply]
Das Thema Strom ist bei mir (noch) schwierig: 
- Ich habe nur vereinzelt Werte aufgezeichnet, da die Ströme für mich bisher nicht von Interesse waren.
- Aus meinen wenigen Werten konnte ich noch nichts sinnvolles ableiten. Evtl. verhält sich Byte 7 & 8 etwas anders als der Rest.
- Mein größtes Problem ist der fehlende Vergleichswert. Entweder ich messe vor Ort selbst oder rechne das über die Leistung der einzelnen Phasen zurück. Meines Wissens nach liefert der VX3 auch keinen Datenpunkt zum Vergleich. 

Ich werde mich demnächst dran machen. 

Ausprobieren kann ich leider nur die Dokumentation der Datenpunkte. Mein Setup ist ein ESP32 mit esphome, da dekodiere ich CAN-MSG selbst.

**Reply by MyHomeMyData** [reply]
@M4n197 Schau Dir mal [diese Diskussion](https://github.com/MyHomeMyData/E3onCAN/discussions/5) an. Denke, die Stromwerte werden auch beim E380 als Int16 Absolutwerte mit 1A Auflösung übertragen. Im Branch develop sind jetzt E380 und E3100CB - hoffentlich - stimmig abgeglichen.

**Reply by M4n197** [reply]
Zu dieser Erkenntnis bin ich auch gekommen.  
Bzgl. des vermuteten Power-Factors innerhalb der letzten zwei Bytes bleibe ich jedoch noch ratlos.

**Comment by MyHomeMyData** [comment]
Bin grade auf Reisen. Schaue mir das nächste Woche an.
-------- Ursprüngliche Nachricht --------Von: wotandamm ***@***.***> Datum: 14.04.24  10:23  (GMT+01:00) An: MyHomeMyData/E3onCAN ***@***.***> Cc: MyHomeMyData ***@***.***>, Mention ***@***.***> Betreff: Re: [MyHomeMyData/E3onCAN] Q&A (Discussion #1) 
super, hinter her ist man immer schlauer ;-).
Baut das jemand ein oder muss die Umwandlung selbst gemacht werden?

**Comment by Gandalf-Avanor** [comment]
@MyHomeMyData: Bei mir läuft seit 1 Monat eine Vitocal 250-A und ein E380 CA-1. 
Heute habe ich versucht, den E380 CA-1 auszulesen und bekomme keine plausiblen Werte:
```
ccea6fe5-open3e:/data/E3onCAN# python3 E3onCANcollect.py -dev e380 -v
2025-11-22 08:55:23.437476 600 GridEnergy: {"ImportCumulated": 8.407790785948902e-45, "ExportCumulated": 4.897538132815236e-45}
2025-11-22 08:55:23.442759 592 GridActivePower: {"L1": -6000.0, "L2": -1.0, "L3": 0.0, "Total": 0.0}
2025-11-22 08:55:23.443242 596 GridCurrent: {"Absolute L1": 50.0, "Absolute L2": 730.0, "Absolute L3": 2.0, "cosPhi": 0.02}
2025-11-22 08:55:23.443510 604 GridEnergy: {"ImportCumulated": 0.0, "Unknown": 0.0}

ccea6fe5-open3e:/data/E3onCAN# python3 E3onCANcollect.py -dev e380 -v -raw
2025-11-22 09:09:41.430101 600 GridEnergy: "701700006f080000"
2025-11-22 09:09:41.435365 592 GridActivePower: "90e8ffff"
2025-11-22 09:09:41.435849 596 GridCurrent: "3200da020200da02"
2025-11-22 09:09:41.436118 604 GridEnergy: "0000"
```
DID 592 und 596 liefern konstant immer die gleichen Werte, die mir nicht plausibel erscheinen.
Dass die DID 600 sowohl bei Import und Export seltsame Werte liefert, kann nicht sein, weil kein Strom exportiert wird. ExportCumulated zeigt schwankende Werte.

Mit CAN-Bus habe ich keinerlei Erfahrung. Wurde der E380 CA-1 evtl. falsch am CAN-Bus angeschlossen (z.B. H und L vertauscht)?

**Reply by MyHomeMyData** [reply]
@Gandalf-Avanor Ich habe eben die aktuelle Version von E3onCAN (V0.4.5) nochmal mit meinem E380 getestet. Das sieht alles plausibel aus. Die neueste Version gibt die Versionsnummer aus, wenn man die Hilfeseite abruft (Option `-h`). Falls das nicht angezeigt wird, bitte auf V0.4.5. updaten. Anleitung steht im [Readme](https://github.com/MyHomeMyData/E3onCAN?tab=readme-ov-file#updating-to-latest-version).

Deine Vitocal scheint ja mit dem E380 klar zu kommen, deshalb dürfte der CAN-Bus kein ernstes Problem haben. Um sicher zu gehen, kannst Du mit `ip -details -statistics link show can0` den Status abrufen und hier posten.

Hab im Moment keine Idee, wie die von Dir beobachteten Effekte zustande kommen könnten.

**Reply by Gandalf-Avanor** [reply]
Die Option -h zeigt, dass ich die aktuelle Version installiert habe:
`E3onCAN V0.4.5 (2025-11-12) using E3 data point definitions as of 20251102`

Der E380 wurde erst später an den CAN-Bus angeschlossen. In ViCare sehe ich keinen Unterschied zu vorher.

Der CAN-Bus ist von 250-A über E380 zum CAN-USB-Adapter verkabelt. Was wäre, wenn CAN-H (von 250-A) auf CAN-L (von E380) auf CAN-H (von Adapter) verkabelt wurde? Wäre dann nur der falsch verdrahtete Knoten gestört oder die gesamte BUS-Kommunikation?
Die Kommunikation von 250-A zu Home Assistant läuft problemlos.

 ```
ccea6fe5-open3e:/data/E3onCAN# ip -details -statistics link show can0
3: can0: <NOARP,UP,LOWER_UP,ECHO> mtu 16 qdisc pfifo_fast state UP mode DEFAULT group default qlen 10
    link/can  promiscuity 0 allmulti 0 minmtu 0 maxmtu 0 
    can state ERROR-ACTIVE restart-ms 0 
	  bitrate 250000 sample-point 0.875
	  tq 250 prop-seg 6 phase-seg1 7 phase-seg2 2 sjw 1 brp 12
	  gs_usb: tseg1 1..16 tseg2 1..8 sjw 1..4 brp 1..1024 brp_inc 1
	  clock 48000000 
	  re-started bus-errors arbit-lost error-warn error-pass bus-off
	  0          0          0          0          0          0         numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535 tso_max_size 65536 tso_max_segs 65535 gro_max_size 65536 gso_ipv4_max_size 65536 gro_ipv4_max_size 65536 parentbus usb parentdev 3-2:1.0 
    RX:  bytes packets errors dropped  missed   mcast           
      33763299 4911403      0       5       0       0 
    TX:  bytes packets errors dropped carrier collsns           
       3164200  395525      0       0       0       0 
```

**Reply by MyHomeMyData** [reply]
Wenn CAN_L und CAN_H vertauscht wären, könnte der CAN2USB-Adapter keine CAN-Botschaften empfangen. Sehr wahrscheinlich würde der Bus zudem so stark gestört, dass die Vitocal CAN-Fehler melden würde.

Auf keinen Fall könnte die Statistik so aussehen wir bei Dir. So sieht es bei einem fehlerfrei funktionierenden Bus aus.

Damit sind wir einer Lösung leider nicht näher gekommen. Mach bitte mal folgendes:
* Diesen Befehl für ca. 10 Sekunden ausführen: `candump -tA canexternal | grep "  258  "`
* Dann für 10 Sekunden diesen: `python3 E3onCANcollect.py -c can0 -dev e380 -v -raw -r 0x258`
* Ausgaben posten.

Falls `candump` nicht funktioniert, zunächst `sudo apt install can-utils` ausführen.

**Reply by Gandalf-Avanor** [reply]
candump:
```
 (2025-11-23 08:00:50.497245)  can0  258   [8]  70 17 00 00 CE 12 00 00
 (2025-11-23 08:00:51.497228)  can0  258   [8]  70 17 00 00 BA 12 00 00
 (2025-11-23 08:00:52.497218)  can0  258   [8]  70 17 00 00 CE 12 00 00
 (2025-11-23 08:00:53.497216)  can0  258   [8]  70 17 00 00 CE 12 00 00
 (2025-11-23 08:00:54.497211)  can0  258   [8]  70 17 00 00 D5 12 00 00
 (2025-11-23 08:00:55.497221)  can0  258   [8]  70 17 00 00 CC 12 00 00
 (2025-11-23 08:00:56.497187)  can0  258   [8]  70 17 00 00 CE 12 00 00
 (2025-11-23 08:00:57.497165)  can0  258   [8]  70 17 00 00 CC 12 00 00
 (2025-11-23 08:00:58.497151)  can0  258   [8]  70 17 00 00 D0 12 00 00
 (2025-11-23 08:00:59.497135)  can0  258   [8]  70 17 00 00 CC 12 00 00
 (2025-11-23 08:01:00.497132)  can0  258   [8]  70 17 00 00 CC 12 00 00
 (2025-11-23 08:01:01.497115)  can0  258   [8]  70 17 00 00 CF 12 00 00
```
E3onCANcollect.py:
```
2025-11-23 09:02:26.496147 600 GridEnergy: "7017000038120000"
2025-11-23 09:02:27.496131 600 GridEnergy: "7017000037120000"
2025-11-23 09:02:28.496137 600 GridEnergy: "7017000040120000"
2025-11-23 09:02:29.496099 600 GridEnergy: "701700003b120000"
2025-11-23 09:02:30.496354 600 GridEnergy: "7017000037120000"
2025-11-23 09:02:31.496093 600 GridEnergy: "7017000040120000"
2025-11-23 09:02:32.496078 600 GridEnergy: "7017000037120000"
2025-11-23 09:02:33.496058 600 GridEnergy: "701700003b120000"
2025-11-23 09:02:34.496063 600 GridEnergy: "7017000037120000"
2025-11-23 09:02:35.496059 600 GridEnergy: "7017000037120000"
2025-11-23 09:02:36.496011 600 GridEnergy: "701700003b120000"
2025-11-23 09:02:37.496007 600 GridEnergy: "701700003b120000"
```

**Reply by MyHomeMyData** [reply]
Tja, ich bin immer noch einigermaßen ratlos. Aus den Daten lese ich:
* Der CAN-Bus funktioniert und liefert plausible Werte. Es treten z.B. keine Sprünge auf.
* E3onCAN übernimmt die Daten korrekt als Raw-Werte. Die Algorithmen zur Dekodierung sind seit über einem Jahr unverändert und liefern mit meinem E380 für sämtliche Datenpunkte vollkommen plausible Werte.
* Der Wert für "ExportCumulated" wird größer **und kleiner**, was bei einem kumulierten Wert ja nicht passieren sollte. 
* Die gewandelten Werte für "ImportCumulated" aus 600 und 604 sind konsistent (0 kWh), aber vermutlich nicht plausibel.
* Die Werte für 592 und 596 sind inkonsistent.

Mir fallen diese möglichen Ursachen ein:
* Dein E380 ist ein anderer Sub-Typ und verwendet andere Datenformate. Falls dem so ist, wäre ich interessiert, die Unterschiede heraus zu bekommen.
* Dein E380 ist CAN-seitig falsch bzw. anders als üblich konfiguriert und verwendet deshalb andere Formate.
* Dein E380 ist Wechselstromseitig falsch angeschlossen (z.B. Reihenfolge der Phasen stimmt nicht).
* Dein E380 hat einen Defekt.

**Reply by MyHomeMyData** [reply]
Da fällt mir noch was ein: Der E380 kann die ganzen Werte ja auch auf dem Display anzeigen. Sind die angezeigten Werte denn konsistent und plausibel?

**Reply by Gandalf-Avanor** [reply]
Bei meiner Internetrecherche zu dem Problem habe ich entdeckt, dass der E380 über ViGuide hinzukonfiguriert werden muss. Das hat mein Heizungsbauer bis jetzt noch nicht gemacht, weshalb ich ihn am Wochenende gleich per Mail dazu aufgefordert habe. Vielleicht wird dabei am E380 eine Konfigurationsänderung vorgenommen.

Auf dem Display kann ich mir nur

- Summe der bezogenen Wirkleistung
- Summe der eingespeisten Wirkleistung
- Summe der bezogenen Blindenergie
- Summe der eingespeisten Blindenergie

jeweils für Tarif 1 und 2 anzeigen lassen. Ich habe nur einen Tarif und die Einspeisung der PV läuft nicht über diesen Zähler.
Die angezeigten Werte sind plausibel. 

Falls mein E380 CA-1 andere Datenformate verwendet, unterstütze ich die Analyse natürlich gerne.

Bei der Installation habe ich den Elektriker gefragt, ob das Drehfeld für die Wärmepumpe relevant ist. Das wusste er nicht. Er sagte jedoch dass er ohne anderslautende Anweisung per Default immer ein rechtsdrehendes Drehfeld installiert.

**Reply by MyHomeMyData** [reply]
> Die angezeigten Werte sind plausibel.

Stimmen die Werte mit den über E3onCAN dekodierten Werten überein?

Der E380CA kann auch die ganzen Einzelwerte anzeigen:
<img width="457" height="780" alt="image" src="https://github.com/user-attachments/assets/2c4e55dd-86fb-432d-9c4e-dca46ea97aa5" />

Falls der E380 nur einphasig angeschlossen ist, würde das die unsinnige Werte für L2 und L3 erklären:
<img width="474" height="305" alt="image" src="https://github.com/user-attachments/assets/b5730b84-d181-4ebc-9f57-2f6301fe7660" />

**Reply by Gandalf-Avanor** [reply]
Der Zähler ist definitiv dreiphasig angeschlossen. Auch die CAN-Bus-Verkabelung habe ich inzwischen überprüft.

Die bei  _**EnErGIES**_ und **_InSt MEASUrES_** angezeigten Werte sind plausibel und stimmig.
Außer bei 0-Werten sehe ich noch nicht einmal eine annähernde Übereinstimmung mit den von E3onCAN ausgegebenen Werten.

Die OK-Taste am Zähler funktioniert nur sporadisch und hat keinen spürbaren Druckpunkt. Habe ich reklamiert.
So langsam glaube ich auch an einen Defekt beim Zähler.

**Reply by Gandalf-Avanor** [reply]
Inzwischen wurde mein E380 CA-1 (wegen der defekten OK-Taste) ausgetauscht. Meine Hoffnung, dass ich damit auch brauchbare Werte über den CAN-Bus auslesen kann, hat sich leider nicht bestätigt.
Der Heizungsbauer hat es auch nicht geschafft, den E380 CA-1 mittels VIGuide einzubinden. Angeblich wäre dazu ein QR-Code auf dem Zähler notwendig gewesen, der jedoch weder auf dem alten noch auf dem neuen Zähler vorhanden ist.
Nachdem der Zähler kostenlos dabei war, habe ich es bei dem unbefriedigenden Zustand belassen und mir einen Shelly Pro 3EM installiert.
